File: 99-recover-binary-search-tree/99-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode* fir;
    TreeNode* prev;
    TreeNode* adj;
    TreeNode* sec;
    public:  
      void recoverTree(TreeNode* root) {
        prev = new TreeNode(INT_MIN);
        inorder(root);
        
        if(fir != NULL && sec != NULL){
            int temp = sec->val;
            sec->val = fir->val;
            fir->val = temp;
        }
        else if(fir != NULL && sec == NULL){
            int temp = fir->val;
            fir->val = adj->val;
            adj->val = temp;
        }
    }
    
    private:
      void inorder(TreeNode* root){
        if(root == NULL) return ;
        
        inorder(root->left);
        if(fir == NULL && root->val < prev->val){
            fir = prev;
            adj = root;
        }
        else if( fir != NULL && root->val < prev->val){
            sec = root;
        }
        prev = root;
        
        inorder(root->right);
    }
};

File: 99-recover-binary-search-tree/99-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> list = new ArrayList<>();
    int idx = 0; 
    
    public void recoverTree(TreeNode root){
        
        inorder(root);
        Collections.sort(list);
        correct(root);
        // System.out.println(list);
        // return root;
    }
    
    private void inorder(TreeNode root){
        if(root == null) return; 
      
        inorder(root.left);
        list.add(root.val);
        inorder(root.right);
    }
    
    private void correct(TreeNode root){
        if(root == null) return;
        
        correct(root.left);
        if(root.val != list.get(idx)){
            root.val = list.get(idx);
        }
        idx++;
        correct(root.right);
    }
}